src/Raven.Client/Documents/BulkInsert/BulkInsertOperation.cs

```diff
@@ -423,14 +517,40 @@ private async Task HandleErrors(string documentId, Exception e)
             await ThrowOnUnavailableStream(documentId, e).ConfigureAwait(false);
         }
 
-        private IDisposable ConcurrencyCheck()
+        private IDisposable _streamLockDisposable;
+        private  ValueTask<ReleaseStream> ConcurrencyCheckAsync()
         {
             if (Interlocked.CompareExchange(ref _concurrentCheck, 1, 0) == 1)
                 throw new InvalidOperationException("Bulk Insert store methods cannot be executed concurrently.");
 
-            return new DisposableAction(() => Interlocked.CompareExchange(ref _concurrentCheck, 0, 1));
+            if (_streamLock.Wait(0))
+            {
+                return new ValueTask<ReleaseStream>(new ReleaseStream(this));
+
+            }
+
+            return new ValueTask<ReleaseStream>(Async());
+
+            async Task<ReleaseStream> Async()
+            {
+                await _streamLock.WaitAsync(_token).ConfigureAwait(false);
+                return new ReleaseStream(this);
+            }
         }
 
+        public readonly struct ReleaseStream : IDisposable
```

Comment:
private, also move to the end of the file, because when doing a merge from 5.4 to 6.0 someone will suffer

Reply:
done